src/promptflow/promptflow/azure/operations/_run_operations.py

```diff
@@ -995,3 +1002,58 @@ def _resolve_flow_definition_resource_id(self, run: Run):
         workspace_id = self._workspace._workspace_id
         location = self._workspace.location
         return f"azureml://locations/{location}/workspaces/{workspace_id}/flows/{run._flow_name}"
+
+    @monitor_operation(activity_name="pfazure.runs.download", activity_type=ActivityType.PUBLICAPI)
+    def download(
+        self, run: Union[str, Run], output: Optional[Union[str, Path]] = None, overwrite: Optional[bool] = False
+    ) -> str:
+        """Download the data of a run, including input, output, snapshot and other run information.
+
+        :param run: The run name or run object
+        :type run: Union[str, ~promptflow.entities.Run]
+        :param output: The output directory. Default to be default to be "~/.promptflow/.runs" folder.
+        :type output: Optional[str]
+        :param overwrite: Whether to overwrite the existing run folder. Default to be False.
+        :type overwrite: Optional[bool]
+        :return: The run directory path
+        :rtype: str
+        """
+        import platform
+
+        run = Run._validate_and_return_run_name(run)
+        if output is None:
+            # default to be "~/.promptflow/.runs" folder
+            output_directory = Path.home() / PROMPT_FLOW_DIR_NAME / PROMPT_FLOW_RUNS_DIR_NAME
+        else:
+            output_directory = Path(output)
+
+        run_folder = output_directory / run
+        if run_folder.exists():
+            if overwrite is True:
+                logger.warning("Removing existing run folder %r.", run_folder.resolve().as_posix())
+                shutil.rmtree(run_folder)
+            else:
+                raise UserErrorException(
+                    f"Run folder {run_folder.resolve().as_posix()!r} already exists, please specify a new output path "
+                    f"or set the overwrite flag to be true."
+                )
+        run_folder.mkdir(parents=True)
+
+        run_downloader = AsyncRunDownloader(run=run, run_ops=self, output_folder=run_folder)
+        if platform.system().lower() == "windows":
+            # Reference: https://stackoverflow.com/questions/45600579/asyncio-event-loop-is-closed-when-getting-loop
+            # On Windows seems to be a problem with EventLoopPolicy, use this snippet to work around it
+            asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
+
+        # when running in notebook, there is already a running event loop, we need to handle this case
+        try:
+            asyncio.get_running_loop()  # Triggers RuntimeError if no running event loop
+            # Create a separate thread so we can block before returning
+            with ThreadPoolExecutor(1) as pool:
+                pool.submit(lambda: asyncio.run(run_downloader.download())).result()
```

Comment:
Will we get correct exception here?

Reply:
Updated to use an existing util func to do this trick, but basically the same logic. Should be fine with the exception I think?

Reply:
I remember directly call result will lose exception sometimes. But since this only runs in 1 thread, I think it's find.